### DISCUSSION TITLE
How do you set a timeout for `pool.Acquire` in pgxpool, so that if the pool is full we can fail fast?

### DISCUSSION THREAD

**Discussion by vamshiaruru** [jackc/pgx#2185] [Q&A] [answered]
Hello, I am using pgxpool (with pgx v5.6.0) for connection pooling. I want to set a timeout on the operation that acquires the connection from the pool, so that if all the connections in the pool are busy for a long period for whatever reason, we fail fast and early. I know that if I set a tracer which satisfies the tracer interface, in the`TraceAcquireStart` function I can create a new context with a timeout and return it. Is that the right way to do this? How do we call the cancel function in that case?
Any help is appreciated, TIA :D

**Comment by jackc** [comment, marked as answer]
If you're directly calling Acquire on the pool then the regular context argument will do what you want. If Acquire is being called implicitly by another function like `Query`, then there is no built-in / intended way to have a separate context passed to Acquire. Off the top of my head, I'm not sure if using a Tracer as you describe would work.

This might actually relate to #2163 and #2171. It just occurred to me that a possible improvement to all of these would be to allow passing a context as a value in another context. e.g. You pass a context to Query, but if that context has a Value with the key of AcquireContext, then that context is used for the Acquire instead of the context passed to Query.

**Reply by jwreschnig-utiq** [reply]
Could it rather be a function returning a context, e.g. `func(queryContext context.Context) (acquireContext context.Context)` rather than a direct context value? That would make it easier to only add the value with a function at e.g. the top level of the program, rather than having to prepare a new context before each query even if we might not need to acquire.

I think this could workaround most of the acquire timeout complications I've run into, including something relatively fancy like preparing an acquire timeout with x% of the time budget of the query context.

**Reply by vamshiaruru** [reply]
Yeah, the acquire is being called implicitly. I use sqlc with pgx and pass the pool object directly to sqlc. I'll explore how to do it with a custom wrapper instead. 
Thanks for the quick reply!

**Reply by vamshiaruru** [reply]
Seems like I have to essentially rewrite a lot of functions that pgxpool implements to get this working. Pgxpool seems to internally use errRow and errRows in QueryRow to return pgx.Rows return values. Do you have any recommendation on the cleanest way to do this?
